fix: cleanup hooks not receiving error signal

felipecrs opened the pull request:
Closes #1041

felipecrs commented:
Any feedback on the approach I took is appreciated. Note that I am only interested on errors thrown in "sync" and "apply". For the other commands, if you believe it would be better to always set err as `nil`, I'm fine too.

yxxhero commented:
@felipecrs could you add some tests for this feature? please ping me if you need any help.

yxxhero commented:
@felipecrs sorry for the ping. any updates? thanks so much.

yxxhero commented:
This pull request includes several changes to the `TriggerCleanupEvent` method in the `pkg/app/app.go` and `pkg/state/state.go` files. The changes primarily focus on modifying the method to accept an additional error parameter (`evtErr`) and updating its calls accordingly.

### Updates to `TriggerCleanupEvent` method:

* [`pkg/app/app.go`](diffhunk://#diff-63facfa8b028c9f397524787fac074737e5534ae2591640c1a029549ec3d7450L1555-R1561): Modified calls to `TriggerCleanupEvent` to pass `nil` or the first error from `errs` as the new `evtErr` parameter. [[1]](diffhunk://#diff-63facfa8b028c9f397524787fac074737e5534ae2591640c1a029549ec3d7450L1555-R1561) [[2]](diffhunk://#diff-63facfa8b028c9f397524787fac074737e5534ae2591640c1a029549ec3d7450L1604-R1610) [[3]](diffhunk://#diff-63facfa8b028c9f397524787fac074737e5534ae2591640c1a029549ec3d7450L1850-R1856)
* [`pkg/state/state.go`](diffhunk://#diff-7e4e87a1e777423a3d9a915e00fb78a9c4fc2f4c476863f60b67e89b9c635697L895-R895): Updated `TriggerCleanupEvent` method to accept an `evtErr` parameter and modified its calls to pass either `relErr` or `nil`. [[1]](diffhunk://#diff-7e4e87a1e777423a3d9a915e00fb78a9c4fc2f4c476863f60b67e89b9c635697L895-R895) [[2]](diffhunk://#diff-7e4e87a1e777423a3d9a915e00fb78a9c4fc2f4c476863f60b67e89b9c635697L1021-R1021) [[3]](diffhunk://#diff-7e4e87a1e777423a3d9a915e00fb78a9c4fc2f4c476863f60b67e89b9c635697L1551-R1556) [[4]](diffhunk://#diff-7e4e87a1e777423a3d9a915e00fb78a9c4fc2f4c476863f60b67e89b9c635697L1652-R1657) [[5]](diffhunk://#diff-7e4e87a1e777423a3d9a915e00fb78a9c4fc2f4c476863f60b67e89b9c635697L1727-R1737) [[6]](diffhunk://#diff-7e4e87a1e777423a3d9a915e00fb78a9c4fc2f4c476863f60b67e89b9c635697L2074-R2084) [[7]](diffhunk://#diff-7e4e87a1e777423a3d9a915e00fb78a9c4fc2f4c476863f60b67e89b9c635697L2428-R2439)
/

felipecrs commented:
@yxxhero the approach I was using was not being effective.

For some reason the `cleanup` hook was being triggered twice, one through the functions I had modified in d77c72f82e771659153746bf1d764889d54c90fb, and another through 12eef5a0eb9029848c2548f7d58bf2f85f1ed6d0.

However, even though both were being called, the first call exits early and doesn't actually get to execute the hook.

With that in mind, I reverted the changes made in d77c72f82e771659153746bf1d764889d54c90fb since I was not able to actually get the hook to be executed through those function calls.

My question is: if the hook doesn't actually get executed through those function calls, I wonder why such function calls exist to begin with.

Anyway, it's now working. I'm testing with the following:

```yaml
hooks:
  - name: global-level
    events:
      - cleanup
    showlogs: true
    command: echo
    args:
      - "error is '{{`{{.Event.Error}}` | not | not}}'"

releases:
  - name: raw
    chart: https://github.com/dysnix/charts/releases/download/raw-v0.3.2/raw-v0.3.2.tgz
    wait: true
    timeout: 5
    values:
      - resources:
          - apiVersion: v1
            kind: Pod
            metadata:
              name: raw
            spec:
              containers:
                - name: raw
                  image: non-existing-image
    hooks:
      - name: release-level
        events:
          - postsync
        showlogs: true
        command: echo
        args:
          - "error is '{{`{{.Event.Error}}` | not | not}}'"
```

Previously:

```console
$ helmfile sync
[...]
hook[postsync] logs | error is 'true'
[...]
hook[cleanup] logs | error is 'false'
[...]
```

Now:

```console
$ helmfile sync
Upgrading release=raw, chart=/home/felipecrs/.cache/helmfile/https_github_com/dysnix/charts/releases/download/raw-v0.3.2/raw-v0.3.2.tgz, namespace=

hook[postsync] logs | error is 'true'
hook[postsync] logs | 

FAILED RELEASES:
NAME   NAMESPACE   CHART                                                                          VERSION   DURATION
raw                https://github.com/dysnix/charts/releases/download/raw-v0.3.2/raw-v0.3.2.tgz                   5s


hook[cleanup] logs | error is 'true'
hook[cleanup] logs | 
in ./helmfile.yaml.gotmpl: failed processing release raw: command "/home/linuxbrew/.linuxbrew/bin/helm" exited with non-zero status:

PATH:
  /home/linuxbrew/.linuxbrew/bin/helm

ARGS:
  0: helm (4 bytes)
  1: upgrade (7 bytes)
  2: --install (9 bytes)
  3: raw (3 bytes)
  4: /home/felipecrs/.cache/helmfile/https_github_com/dysnix/charts/releases/download/raw-v0.3.2/raw-v0.3.2.tgz (106 bytes)
  5: --wait (6 bytes)
  6: --timeout (9 bytes)
  7: 5s (2 bytes)
  8: --create-namespace (18 bytes)
  9: --values (8 bytes)
  10: /tmp/helmfile1744551334/raw-values-869f5446cb (45 bytes)
  11: --reset-values (14 bytes)
  12: --history-max (13 bytes)
  13: 10 (2 bytes)

ERROR:
  exit status 1

EXIT STATUS
  1

STDERR:
  Error: UPGRADE FAILED: context deadline exceeded

COMBINED OUTPUT:
  Error: UPGRADE FAILED: context deadline exceeded
```

felipecrs commented:
> @felipecrs could you add some tests for this feature? please ping me if you need any help.

Can you help me with that? I am modifying `run.go`, and there are no existing tests for this file.

yxxhero commented:
@felipecrs I will help you.

yxxhero commented:
see: https://github.com/helmfile/helmfile/pull/2475
